test/SparqlAntlrParserTestHelpers.h

```diff
@@ -122,279 +193,489 @@ constexpr const ad_utility::Last<Current, Others...>* unwrapVariant(
-  };
-  ASSERT_EQ(resultOfParseAndText.resultOfParse_.size(), sizeof...(matchers));
-  auto sequence = std::make_index_sequence<sizeof...(matchers)>();
+namespace variant_matcher {
+// Implements a matcher that checks the value of arbitrarily deeply nested
+// variants that contain a value with the last Type of the Ts. The variant may
+// also be an suffix of the Ts so to speak. E.g.
+// `MultiVariantMatcher<VarOrTerm, GraphTerm, Iri>` can match on a `Iri`,
+// `GraphTerm=std::variant<Literal, BlankNode, Iri>` and
+// `VarOrTerm=std::variant<Variable, GraphTerm>`.
+template <typename... Ts>
+class MultiVariantMatcher {
+ public:
+  explicit MultiVariantMatcher(
+      ::testing::Matcher<const ad_utility::Last<Ts...>&> matcher)
+      : matcher_(std::move(matcher)) {}
+
+  template <typename Variant>
+  bool MatchAndExplain(const Variant& value,
+                       ::testing::MatchResultListener* listener) const {
+    auto elem = unwrapVariant<Ts...>(value);
+    if (!listener->IsInterested()) {
+      return elem && matcher_.Matches(*elem);
+    }
 
-  auto lambda = [&]<size_t... i>(std::index_sequence<i...>) {
-    (...,
-     expect_single_element(resultOfParseAndText.resultOfParse_[i], matchers));
-  };
+    if (!elem) {
+      *listener << "whose value is not of type '" << GetTypeName() << "'";
+      return false;
+    }
 
-  lambda(sequence);
-  EXPECT_TRUE(resultOfParseAndText.remainingText_.empty());
-}
-// _____________________________________________________________________________
+    testing::StringMatchResultListener elem_listener;
+    const bool match = matcher_.MatchAndExplain(*elem, &elem_listener);
+    *listener << "whose value " << testing::PrintToString(*elem)
+              << (match ? " matches" : " doesn't match");
+    // First add our own Explanation and then that of the sub matcher.
+    auto subMatcherExplanation = elem_listener.str();
+    if (!subMatcherExplanation.empty()) {
+      *listener << subMatcherExplanation;
+    }
+    return match;
+  }
 
-MATCHER_P(IsIri, value, "") {
-  if (const auto iri = unwrapVariant<VarOrTerm, GraphTerm, Iri>(arg)) {
-    return iri->iri() == value;
+  void DescribeTo(std::ostream* os) const {
+    *os << "is a variant<> with value of type '" << GetTypeName()
+        << "' and the value ";
+    matcher_.DescribeTo(os);
   }
-  return false;
-}
 
-// _____________________________________________________________________________
+  void DescribeNegationTo(std::ostream* os) const {
+    *os << "is a variant<> with value of type other than '" << GetTypeName()
+        << "' or the value ";
+    matcher_.DescribeNegationTo(os);
+  }
 
-MATCHER_P2(IsBlankNode, generated, label, "") {
-  if (const auto blankNode =
-          unwrapVariant<VarOrTerm, GraphTerm, BlankNode>(arg)) {
-    return blankNode->isGenerated() == generated && blankNode->label() == label;
+ private:
+  static std::string GetTypeName() {
+    return testing::internal::GetTypeName<ad_utility::Last<Ts...>>();
```

Comment:
This is also an internal gtest function. But getting a type as a string in c++ seems to non trivial. `testing::internal::GetTypeName<...>()` does somewhat lowlevel stuff. I don't whether it is worth reimplementing the functionality to get rid of this.